openverse_catalog/dags/data_refresh/dag_factory.py

```diff
@@ -0,0 +1,186 @@
+ensure that no two data refresh DAGs can run concurrently, as required by
+the server.
+
+A data refresh occurs on the data refresh server in the openverse-api project.
+This is a task which imports data from the upstream Catalog database into the
+API, copies contents to a new Elasticsearch index, and makes the index "live".
+This process is necessary to make new content added to the Catalog by our
+provider DAGs available on the frontend.
+
+The DAGs generated by this factory allow us to trigger those refreshes through
+Airflow. Since no two refreshes can run simultaneously, all tasks are run in a
+special `data_refresh` pool with a single worker slot. To ensure that tasks
+run in an acceptable order (ie the trigger step for one DAG cannot run if a
+previously triggered refresh is still running), each DAG has the following
+steps:
+
+1. The `wait_for_data_refresh` step uses a custom Sensor that will wait until
+none of the `external_dag_ids` (corresponding to the other data refresh DAGs)
+are 'running'. A DAG is considered to be 'running' if it is itself in the
+RUNNING state __and__ its own `wait_for_data_refresh` step has completed
+successfully. The Sensor suspends itself and frees up the worker slot if
+another data refresh DAG is running.
+
+2. The `trigger_data_refresh` step then triggers the data refresh by POSTing
+to the `/task` endpoint on the data refresh server with relevant data. A
+successful response will include the `status_check` url used to check on the
+status of the refresh, which is passed on to the next task via XCom.
+
+3. Finally the `wait_for_data_refresh` task waits for the data refresh to be
+complete by polling the `status_url`. Note this task does not need to be
+able to suspend itself and free the worker slot, because we want to lock the
+entire pool on waiting for a particular data refresh to run.
+
+You can find more background information on this process in the following
+issues and related PRs:
+
+- [[Feature] Data refresh orchestration DAG](
+https://github.com/WordPress/openverse-catalog/issues/353)
+"""
+# airflow DAG (necessary for Airflow to find this file)
+import json
+import logging
+from typing import Sequence
+from urllib.parse import urlparse
+
+from airflow import DAG
+from airflow.exceptions import AirflowException
+from airflow.providers.http.operators.http import SimpleHttpOperator
+from airflow.providers.http.sensors.http import HttpSensor
+from common.sensors import ExternalDAGsSensor
+from data_refresh.data_refresh_types import DATA_REFRESH_CONFIGS, DataRefresh
+
+
+logger = logging.getLogger(__name__)
+
+
+XCOM_PULL_TEMPLATE = "{{{{ ti.xcom_pull(task_ids='{}', key='{}') }}}}"
+DATA_REFRESH_POOL = "data_refresh"
+
+
+# Filter for the `trigger_data_refresh` task, used to grab the endpoint needed
+# to poll for the status of the triggered data refresh. This information will
+# then be available via XCom in the downstream tasks.
+def response_filter_data_refresh(response):
```

Comment:
Sorry, I might have been unclear - I think this is a perfect docstring:
```suggestion
def response_filter_data_refresh(response):
    """
    Filter for the `trigger_data_refresh` task, used to grab the endpoint needed
    to poll for the status of the triggered data refresh. This information will
    then be available via XCom in the downstream tasks.
    """
```

Same for the function below! Type annotations would also be useful on these two functions.

Reply:
Ah yep-- just missed these as I was moving comments around. These should definitely be docstrings, will do.